pkg/controller/sparkapplication/controller.go

```diff
@@ -507,15 +501,17 @@ func (c *Controller) syncSparkApplication(key string) error {
 				appToUpdate.Namespace, appToUpdate.Name, err)
 			return err
 		}
+		c.clearStatusWhenInvalidated(&appToUpdate.Status)
 		appToUpdate.Status.AppState.State = v1beta1.PendingRerunState
-		appToUpdate.Status.ExecutionAttempts = 0
 	case v1beta1.PendingRerunState:
 		if c.validateSparkResourceDeletion(appToUpdate) {
-			// Reset SubmissionAttempts count since this is a new overall run.
-			appToUpdate.Status.SubmissionAttempts = 0
-			appToUpdate.Status.TerminationTime = metav1.Time{}
+			c.recordSparkApplicationEvent(appToUpdate)
```

Comment:
Probably it would be good to run here `clearStatusWhenInvalidated` and `deleteSparkResources` to make sure just before each submission that we start fresh.

Reply:
When the state is `PendingRerun`, we don't need to reset the status because the application is immediately submitted in that case and status is reset when the application gets submitted.

Reply:
I meant executor status. It seems that this is not enough as executor state is carried because exec pods left around from previous runs trigger updates when the app starts running.
Check the [log](https://gist.github.com/skonto/3705de9b88aa4830dac9506662fa82c3) for example these lines:
```
I0318 10:45:16.152909      11 event.go:221] Event(v1.ObjectReference{Kind:"SparkApplication", Namespace:"spark", Name:"spark-pi", UID:"c9c55684-496a-11e9-91f6-0aee7538e092", APIVersion:"sparkoperator.k8s.io/v1beta1", ResourceVersion:"2196261", FieldPath:""}): type: 'Warning' reason: 'SparkExecutorFailed' Executor spark-pi-1552905893644-exec-1 failed
I0318 10:45:16.156769      11 controller.go:521] Trying to update SparkApplication spark/spark-pi, from: [{ 2019-03-18 10:45:16 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 30545    } {SUBMITTED } map[] 1 1}] to [{spark-application-1552905905083 2019-03-18 10:45:16 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 30545    spark-pi-driver} {SUBMITTED } map[spark-pi-1552905893644-exec-1:FAILED] 1 1}]
I0318 10:45:16.165868      11 sparkapp_metrics.go:144] Exporting metrics for spark-pi; old status: { 2019-03-18 10:45:16 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 30545    } {SUBMITTED } map[] 1 1} new status: {spark-application-1552905905083 2019-03-18 10:45:16 +0000 UTC 0001-01-01 00:00:00 +0000 UTC {spark-pi-ui-svc 30545    spark-pi-driver} {SUBMITTED } map[spark-pi-1552905893644-exec-1:FAILED] 1 1}
I0318 10:45:16.165917      11 sparkapp_metrics.go:221] Exporting Metrics for Executor spark-pi-1552905893644-exec-1. OldState:  NewState: FAILED
I0318 10:45:16.165937      11 metrics.go:124] Decrementing spark_app_executor_running_count with labels map[app_type:Unknown] metricVal to 1
I0318 10:45:16.165955      11 controller.go:253] Ending processing key: "spark/spark-pi"
```
The old executor at some point triggers a pod update which triggers an app update, although a few lines above we have a new submit. State is different and syncSparkApplication make it become part of the app exec status and is never removed.  

Now regarding the event record. I added some logging in `getAndUpdateAppState`.

In the same scenario different run:
```
  Executor State:
    Spark - Pi - 1552909628546 - Exec - 2:  FAILED
    Spark - Pi - 1552909656575 - Exec - 1:  COMPLETED
    Spark - Pi - 1552909656575 - Exec - 2:  RUNNING
  Last Submission Attempt Time:             2019-03-18T11:47:38Z
  Spark Application Id:                     spark-application-1552909673112
  Submission Attempts:                      1
  Termination Time:                         2019-03-18T11:49:51Z
Events:
  Type     Reason                               Age                    From            Message
  ----     ------                               ----                   ----            -------
  Normal   SparkApplicationAdded                5m15s                  spark-operator  SparkApplication spark-pi was added, enqueuing it for submission
  Normal   SparkExecutorPending                 5m2s (x2 over 5m2s)    spark-operator  Executor spark-pi-1552909576599-exec-1 is pending
  Normal   SparkExecutorPending                 5m2s                   spark-operator  Executor spark-pi-1552909576599-exec-2 is pending
  Normal   SparkExecutorRunning                 4m57s                  spark-operator  Executor spark-pi-1552909576599-exec-2 is running
  Normal   SparkExecutorRunning                 4m57s                  spark-operator  Executor spark-pi-1552909576599-exec-1 is running
  Warning  SparkExecutorFailed                  4m41s                  spark-operator  Executor spark-pi-1552909576599-exec-1 failed
  Warning  SparkExecutorFailed                  4m41s                  spark-operator  Executor spark-pi-1552909576599-exec-2 failed
  Normal   SparkExecutorPending                 4m26s (x2 over 4m26s)  spark-operator  Executor spark-pi-1552909607670-exec-1 is pending
  Normal   SparkApplicationSpecUpdateProcessed  4m24s (x2 over 4m44s)  spark-operator  Successfully processed spec update for SparkApplication spark-pi
  Warning  SparkApplicationPendingRerun         4m23s (x2 over 4m44s)  spark-operator  SparkApplication spark-pi is pending rerun
  Normal   SparkApplicationSubmitted            4m20s (x3 over 5m12s)  spark-operator  SparkApplication spark-pi was submitted successfully
  Normal   SparkExecutorPending                 4m20s (x3 over 4m26s)  spark-operator  Executor spark-pi-1552909607670-exec-2 is pending
  Normal   SparkDriverRunning                   4m14s (x3 over 5m6s)   spark-operator  Driver spark-pi-driver is running
  Normal   SparkExecutorPending                 4m11s (x2 over 4m11s)  spark-operator  Executor spark-pi-1552909628546-exec-1 is pending
```
Looking at the logs: 
```
I0318 11:47:19.705428      11 spark_pod_eventhandler.go:41] Pod spark-pi-1552909628546-exec-2 added in namespace spark.
I0318 11:47:19.705460      11 spark_pod_eventhandler.go:77] Enqueuing SparkApplication spark/spark-pi for app update processing.
I0318 11:47:19.705490      11 controller.go:246] Starting processing key: "spark/spark-pi"
I0318 11:47:19.705552      11 controller.go:300] getAndUpdateAppState spark-pi-1552909628546-exec-1 2019-03-18 11:47:19.705544594 +0000 UTC
I0318 11:47:19.705588      11 controller.go:300] getAndUpdateAppState spark-pi-1552909628546-exec-2 2019-03-18 11:47:19.705582392 +0000 UTC
I0318 11:47:19.705607      11 controller.go:305] getAndUpdateAppState record event spark-pi-1552909628546-exec-2 2019-03-18 11:47:19.705582392 +0000 UTC
```
The recording there is never shown in the description, weird.

Reply:
>  I think options are either block updates for old pods in getAndUpdateAppState or make sure all resources are removed before a new submission happens (probably too restrictive).

I think we need to have the concept of some submission ID that is unique to each submission, and then use it when listing pods or processing pod events. This submission ID gets added to the pods as a label.

Reply:
> The recording there for spark-pi-1552909628546-exec-2 is never shown in the event list weird. Shouldn't all the events be listed there?

Yes, this is weird. I cannot explain this given that the code does record events for the new executors.

Reply:
> I think we need to have the concept of some submission ID that is unique to each submission, and then use it when listing pods or processing pod events. This submission ID gets added to the pods as a label.

That also will work assuming we know which is the current submission id when an a random update occurs.

Reply:
Yes, we will record the submission ID in `SparkApplicationStatus`.

Reply:
I added a call to `clearStatus` here also. PTAL.

Reply:
@liyinan926 it seems that it overcomes the issue with the updates:
```
Executor State:
    Spark - Pi - 1552944088510 - Exec - 1:  RUNNING
    Spark - Pi - 1552944088510 - Exec - 2:  COMPLETED
  Last Submission Attempt Time:             2019-03-18T21:21:30Z
  Spark Application Id:                     spark-application-1552944104139
  Submission Attempts:                      1
  Termination Time:                         2019-03-18T21:25:25Z
Events:
  Type     Reason                               Age                    From            Message
  ----     ------                               ----                   ----            -------
  Normal   SparkApplicationAdded                6m9s (x2 over 6m12s)   spark-operator  SparkApplication spark-pi was added, enqueuing it for submission
  Normal   SparkExecutorPending                 5m54s (x2 over 5m54s)  spark-operator  Executor spark-pi-1552944036721-exec-2 is pending
  Normal   SparkExecutorPending                 5m54s (x2 over 5m54s)  spark-operator  Executor spark-pi-1552944036721-exec-1 is pending
  Normal   SparkExecutorRunning                 5m49s                  spark-operator  Executor spark-pi-1552944036721-exec-1 is running
  Normal   SparkExecutorRunning                 5m49s                  spark-operator  Executor spark-pi-1552944036721-exec-2 is running
  Warning  SparkExecutorFailed                  5m38s                  spark-operator  Executor spark-pi-1552944036721-exec-2 failed
  Normal   SparkExecutorPending                 5m24s (x2 over 5m24s)  spark-operator  Executor spark-pi-1552944067633-exec-2 is pending
  Normal   SparkExecutorPending                 5m24s (x2 over 5m24s)  spark-operator  Executor spark-pi-1552944067633-exec-1 is pending
  Normal   SparkApplicationSpecUpdateProcessed  5m20s (x2 over 5m41s)  spark-operator  Successfully processed spec update for SparkApplication spark-pi
  Warning  SparkApplicationPendingRerun         5m20s (x2 over 5m41s)  spark-operator  SparkApplication spark-pi is pending rerun
  Normal   SparkApplicationSubmitted            5m17s (x3 over 6m9s)   spark-operator  SparkApplication spark-pi was submitted successfully
  Normal   SparkDriverRunning                   5m13s (x3 over 6m3s)   spark-operator  Driver spark-pi-driver is running
  Normal   SparkExecutorPending                 5m3s (x2 over 5m3s)    spark-operator  Executor spark-pi-1552944088510-exec-1 is pending
```
feel free to commit this one.
The weird thing is that only the `spark-pi-1552944088510-exec-1` events are shown although 
`spark-pi-1552944088510-exec-2` als was recorded:
```
I0318 21:21:44.521557      11 spark_pod_eventhandler.go:41] Pod spark-pi-1552944088510-exec-2 added in namespace spark.
I0318 21:21:44.521582      11 spark_pod_eventhandler.go:77] Enqueuing SparkApplication spark/spark-pi for app update processing.
I0318 21:21:44.521609      11 controller.go:246] Starting processing key: "spark/spark-pi"
I0318 21:21:44.521664      11 controller.go:298] getAndUpdateAppState spark-pi-1552944088510-exec-2 2019-03-18 21:21:44.52165929 +0000 UTC
I0318 21:21:44.521685      11 controller.go:304] getAndUpdateAppState record event spark-pi-1552944088510-exec-2 2019-03-18 21:21:44.52165929 +0000 UTC
I0318 21:21:44.521693      11 controller.go:846] Blabla: PENDING 2019-03-18 21:21:44.521690233 +0000 UTC
I0318 21:21:44.521723      11 controller.go:298] getAndUpdateAppState spark-pi-1552944088510-exec-1 2019-03-18 21:21:44.521717418 +0000 UTC
I0318 21:21:44.521756      11 event.go:221] Event(v1.ObjectReference{Kind:"SparkApplication", Namespace:"spark", Name:"spark-pi", UID:"ab428214-49c3-11e9-91f6-0aee7538e092", APIVersion:"sparkoperator.k8s.io/v1beta1", ResourceVersion:"2323655", FieldPath:""}): type: 'Normal' reason: 'SparkExecutorPending' Executor spark-pi-1552944088510-exec-2 is pending
I0318 21:21:44.526176      11 spark_pod_eventhandler.go:52] Pod spark-pi-1552944088510-exec-2 updated in namespace spark.
```

Reply:
Sounds good. Thanks!